File: packet_seaquence/log/go_src/analyzer/analyzer.go
// Package analyzer provides the main packet sequence analysis functionality
package analyzer

import (
	"fmt"
	"os"
	
	"app/go_src/extractor"
	"app/go_src/models"
	"app/go_src/protocol_analyzer"
	"app/go_src/reader"
	"app/go_src/writer"
)

// PacketSequenceAnalyzer provides functionality to analyze packet sequences
type PacketSequenceAnalyzer struct {
	PcapFile    string
	OutputFile  string
	MaxEntries  int
	Options     map[string]interface{}
}

// PacketAdapter adapts reader.Packet to protocol_analyzer.Packet
type PacketAdapter struct {
	packet *reader.Packet
}

// LayerExists implements protocol_analyzer.Packet.LayerExists
func (pa *PacketAdapter) LayerExists(layerType string) bool {
	return pa.packet.LayerExists(layerType)
}

// GetField implements protocol_analyzer.Packet.GetField
func (pa *PacketAdapter) GetField(layer string, field string) (interface{}, bool) {
	return pa.packet.GetField(layer, field)
}

// Layer implements protocol_analyzer.Packet.Layer
func (pa *PacketAdapter) Layer(layerType string) interface{} {
	// For tshark-based implementation, we'll check if the layer exists and return the Info map
	// for that layer if it does
	if !pa.packet.LayerExists(layerType) {
		return nil
	}
	
	// Return the layer info if available
	if info, ok := pa.packet.Info[layerType]; ok {
		return info
	}
	
	// Otherwise just return an empty map to indicate the layer exists but has no details
	return map[string]interface{}{}
}

// GetProtocolName implements protocol_analyzer.Packet.GetProtocolName
func (pa *PacketAdapter) GetProtocolName() string {
	return pa.packet.GetProtocolName()
}

// GetTimestamp returns the packet timestamp
func (pa *PacketAdapter) GetTimestamp() string {
	return pa.packet.Timestamp.String()
}

// GetSrcIP returns the source IP
func (pa *PacketAdapter) GetSrcIP() string {
	return pa.packet.SrcIP
}

// GetDstIP returns the destination IP
func (pa *PacketAdapter) GetDstIP() string {
	return pa.packet.DstIP
}

// NewPacketSequenceAnalyzer creates a new PacketSequenceAnalyzer
func NewPacketSequenceAnalyzer(pcapFile, outputFile string, maxEntries int, options map[string]interface{}) (*PacketSequenceAnalyzer, error) {
	// Check if file exists
	if _, err := os.Stat(pcapFile); os.IsNotExist(err) {
		return nil, fmt.Errorf("PCAPファイルが見つかりません: %s", pcapFile)
	}
	
	// If options is nil, initialize it to an empty map
	if options == nil {
		options = make(map[string]interface{})
	}
	
	return &PacketSequenceAnalyzer{
		PcapFile:    pcapFile,
		OutputFile:  outputFile,
		MaxEntries:  maxEntries,
		Options:     options,
	}, nil
}

// convertPackets converts []reader.Packet to []protocol_analyzer.Packet
func convertPackets(readerPackets []reader.Packet) []protocol_analyzer.Packet {
	protocolPackets := make([]protocol_analyzer.Packet, len(readerPackets))
	for i := range readerPackets {
		 // Create a new adapter for each packet
		protocolPackets[i] = &PacketAdapter{packet: &readerPackets[i]}
	}
	return protocolPackets
}

// Analyze performs the packet sequence analysis
func (a *PacketSequenceAnalyzer) Analyze() error {
	// 1. Read the PCAP file
	pcapReader := reader.NewPCAPReader(a.PcapFile)
	err := pcapReader.Open()
	if err != nil {
		return fmt.Errorf("PCAPファイルオープン中にエラーが発生しました: %w", err)
	}
	defer pcapReader.Close()
	
	// verbose flag from options
	verbose := false
	if v, ok := a.Options["verbose"].(bool); ok {
		verbose = v
	}
	
	// Read all packets (maxPackets=0 means no limit)
	packets, err := pcapReader.Read(0, verbose)
	if err != nil {
		return fmt.Errorf("PCAPファイル読み込み中にエラーが発生しました: %w", err)
	}
	
	// 2. Extract data from the packets
	// Convert packets to protocol_analyzer.Packet type
	protocolPackets := convertPackets(packets)
	dataExtractor := extractor.NewExtractData(protocolPackets, len(protocolPackets)) // Extract all packets
	packetData := dataExtractor.Extract()
	
	// Debug output if verbose is enabled
	if verbose, ok := a.Options["verbose"].(bool); ok && verbose {
		a.debugPacketSequenceData(packetData)
	}
	
	// 3. Get all packets and divide them into chunks
	allPackets := packetData.GetPackets()
	totalPackets := len(allPackets)
	
	// Calculate chunk size
	chunkSize := a.MaxEntries
	if chunkSize <= 0 {
		chunkSize = totalPackets
	}
	
	// Divide packets into chunks
	chunks := make([]*models.PacketSequenceData, 0)
	for i := 0; i < totalPackets; i += chunkSize {
		// Calculate end index for this chunk
		end := i + chunkSize
		if end > totalPackets {
			end = totalPackets
		}
		
		// Create a new PacketSequenceData for this chunk
		chunkData := models.NewPacketSequenceData()
		for j := i; j < end; j++ {
			chunkData.AddPacket(allPackets[j])
		}
		
		chunks = append(chunks, chunkData)
	}
	
	// 4. Generate Mermaid diagrams for each chunk
	mermaidContents := make([]string, 0, len(chunks))
	for _, chunkData := range chunks {
		mermaidWriter := writer.NewWriteMermaid(chunkData)
		mermaidContent := mermaidWriter.Generate()
		mermaidContents = append(mermaidContents, mermaidContent)
	}
	
	// 5. Write the Mermaid diagrams to files
	fileWriter := writer.NewWriteFile(a.OutputFile)
	
	if len(mermaidContents) == 1 {
		// If there's only one chunk, write it directly
		if err := fileWriter.Write(mermaidContents[0]); err != nil {
			return fmt.Errorf("ファイル出力中にエラーが発生しました: %w", err)
		}
		fmt.Printf("解析が完了しました。出力ファイル: %s\n", a.OutputFile)
	} else {
		// If there are multiple chunks, write them to separate files
		outputFiles, err := fileWriter.WriteSplit(mermaidContents)
		if err != nil {
			return fmt.Errorf("ファイル出力中にエラーが発生しました: %w", err)
		}
		fmt.Printf("解析が完了しました。%d個のファイルに分割出力しました。\n", len(outputFiles))
		for i, file := range outputFiles {
			fmt.Printf("  %d. %s\n", i+1, file)
		}
	}
	
	return nil
}

// debugPacketSequenceData outputs debug information about the PacketSequenceData
func (a *PacketSequenceAnalyzer) debugPacketSequenceData(data *models.PacketSequenceData) {
	fmt.Println("\n===== PacketSequenceData DEBUG =====")
	
	// Get packets and display basic info
	packets := data.GetPackets()
	fmt.Printf("Total packets: %d\n", len(packets))
	
	// Display details of the first few packets
	maxPacketsToShow := 3
	if len(packets) < maxPacketsToShow {
		maxPacketsToShow = len(packets)
	}
	
	for i := 0; i < maxPacketsToShow; i++ {
		packet := packets[i]
		fmt.Printf("\nPacket %d:\n", i+1)
		fmt.Printf("  Protocol: %s\n", packet.Protocol)
		fmt.Printf("  ProtocolNameFromPacket: %s\n", packet.ProtocolNameFromPacket)
		fmt.Printf("  Src: %s\n", packet.Source)
		fmt.Printf("  Dst: %s\n", packet.Dest)
		fmt.Printf("  Time: %v\n", packet.Time)
		
		// Display info keys
		infoKeys := make([]string, 0, len(packet.Info))
		for key := range packet.Info {
			infoKeys = append(infoKeys, key)
		}
		fmt.Printf("  Info keys: %v\n", infoKeys)
		
		// Display protocol-specific info
		for key, value := range packet.Info {
			fmt.Printf("  %s: %v\n", key, value)
		}
	}
	
	fmt.Println("=================================\n")
	
	// Debug the WriteMermaid._build_message method
	fmt.Println("\n===== WriteMermaid DEBUG =====")
	mermaidWriter := writer.NewWriteMermaid(models.NewPacketSequenceData())
	
	for i := 0; i < maxPacketsToShow; i++ {
		if i >= len(packets) {
			break
		}
		
		packet := packets[i]
		message := mermaidWriter.BuildMessage(packet)
		fmt.Printf("Packet %d message: '%s'\n", i+1, message)
	}
	
	fmt.Println("=================================\n")
}

File: packet_seaquence/log/python_no_shark/src/analyzer.py
#!/usr/bin/env python3
"""
PacketSequenceAnalyzerクラスを提供するモジュール
"""
import os
from .reader import ReadPcap
from .extractor import ExtractData
from .writer.mermaid import WriteMermaid
from .writer.file import WriteFile
from .models import PacketSequenceData


class PacketSequenceAnalyzer:
    """
    パケットシーケンス解析の全体の流れを管理するクラス
    
    設計書に記載されている通り、以下の流れで処理を行う:
    1. PCAPファイルの読み込み
    2. データの抽出
    3. Mermaid記法によるシーケンス図の生成
    4. ファイル出力
    """
    
    def __init__(self, pcap_file, output_file="output.mmd", max_entries=50, options=None):
        """
        PacketSequenceAnalyzerクラスの初期化
        
        Args:
            pcap_file (str): PCAPファイルのパス
            output_file (str): 出力ファイルのパス (デフォルト: output.mmd)
            max_entries (int): 抽出するエントリの最大数 (デフォルト: 50)
            options (dict): オプション設定
        """
        self.pcap_file = pcap_file
        self.output_file = output_file
        self.max_entries = max_entries
        self.options = options if options else {}
        
        # ファイルの存在確認
        if not os.path.exists(pcap_file):
            raise FileNotFoundError(f"PCAPファイルが見つかりません: {pcap_file}")
    
    def analyze(self):
        """
        パケットシーケンス解析を実行する
        
        処理の流れ:
        1. PCAPファイルを読み込む
        2. パケットからデータを抽出する
        3. Mermaid記法でシーケンス図を生成する
        4. 最大パケット数に基づいてファイルに分割出力する
        
        Returns:
            bool: 解析が成功したかどうか
        """
        try:
            # PCAPファイルの読み込み
            reader = ReadPcap(self.pcap_file)
            packets = reader.read()
            
            # データの抽出
            extractor = ExtractData(packets, len(packets))  # 全パケットを抽出
            packet_data = extractor.extract()
            
            # PacketSequenceDataオブジェクトの詳細なデバッグ出力
            if 'verbose' in self.options and self.options['verbose']:
                print("\n===== PacketSequenceData DEBUG =====")
                print(f"Type: {type(packet_data)}")
                
                # パケット一覧を取得して内容をデバッグ表示
                packets_list = packet_data.get_packets()
                print(f"Total packets: {len(packets_list)}")
                
                # 最初の数パケットの詳細を表示
                for i, packet in enumerate(packets_list[:3]):  # 最初の3パケットだけ表示
                    print(f"\nPacket {i+1}:")
                    print(f"  Protocol: {packet.get('protocol', 'UNKNOWN')}")
                    print(f"  Src: {packet.get('src', 'Unknown')}")
                    print(f"  Dst: {packet.get('dst', 'Unknown')}")
                    print(f"  Time: {packet.get('time', 'Unknown')}")
                    
                    # infoの内容を表示
                    info = packet.get('info', {})
                    print(f"  Info keys: {list(info.keys())}")
                    
                    # 各プロトコル情報の詳細を表示
                    for key in info:
                        print(f"  {key}: {info[key]}")
                
                print("=================================\n")
                
                # WriteMermaidのbuild_messageメソッドをデバッグ
                print("\n===== WriteMermaid DEBUG =====")
                mermaid_writer = WriteMermaid(packet_data.copy_first_n(3))
                # 最初の数パケットでbuild_messageをテスト
                packets_list = packet_data.get_packets()
                for i, packet in enumerate(packets_list[:3]):
                    message = mermaid_writer._build_message(packet)
                    print(f"Packet {i+1} message: '{message}'")
                print("=================================\n")
            
            # パケットリストを取得
            all_packets = packet_data.get_packets()
            total_packets = len(all_packets)
            
            # パケット数をmax_entriesで分割
            chunk_size = self.max_entries
            chunks = []
            
            for i in range(0, total_packets, chunk_size):
                chunk = all_packets[i:i+chunk_size]
                # PacketSequenceDataオブジェクトを正しくインポートされた状態で作成
                chunk_data = PacketSequenceData()
                for packet in chunk:
                    chunk_data.add_packet(packet)
                chunks.append(chunk_data)
            
            # 各チャンクに対してMermaid図を生成
            mermaid_contents = []
            for chunk_data in chunks:
                mermaid_writer = WriteMermaid(chunk_data)
                mermaid_content = mermaid_writer.generate()
                mermaid_contents.append(mermaid_content)
            
            # ファイルに出力
            file_writer = WriteFile(self.output_file)
            
            # 出力が1つだけならそのまま書き込み、複数なら分割書き込み
            if len(mermaid_contents) == 1:
                file_writer.write(mermaid_contents[0])
                print(f"解析が完了しました。出力ファイル: {self.output_file}")
            else:
                file_writer.write_split(mermaid_contents)
                print(f"解析が完了しました。{len(mermaid_contents)}個のファイルに分割出力しました。")
            
            return True
            
        except Exception as e:
            print(f"解析中にエラーが発生しました: {str(e)}")
            import traceback
            traceback.print_exc()
            return False
